source/dub/dependencyresolver.d

```diff
@@ -54,236 +54,149 @@ class DependencyResolver(CONFIGS, CONFIG) {
 	{
-		auto root_base_pack = basePackage(root.pack);
+		import std.bitmanip : BitArray;
 
-		// find all possible configurations of each possible dependency
-		size_t[string] package_indices;
-		string[size_t] package_names;
-		PackageConfigs[] configs;
-		bool[string] maybe_optional_deps;
-		bool[TreeNode] visited;
+		// Leave the possibility to opt-out from the loop limit
+		import std.process : environment;
+		bool no_loop_limit = environment.get("DUB_NO_RESOLVE_LIMIT") !is null;
 
-		void findConfigsRec(TreeNode parent, bool parent_unique)
-		{
-			if (parent in visited) return;
-			visited[parent] = true;
-
-			foreach (ch; getChildren(parent)) {
-				auto basepack = basePackage(ch.pack);
-				auto pidx = configs.length;
-
-				if (ch.depType != DependencyType.required) maybe_optional_deps[ch.pack] = true;
-
-				PackageConfigs config;
-				if (auto pi = basepack in package_indices) {
-					pidx = *pi;
-					config = configs[*pi];
-				} else {
-					if (basepack == root_base_pack) config.allConfigs = [root.config];
-					else config.allConfigs = getAllConfigs(basepack);
-					configs ~= config;
-					package_indices[basepack] = pidx;
-					package_names[pidx] = basepack;
+		// build up the dependency graph, eliminating as many configurations/
+		// versions as possible
+		PackageConfigs[string] packs;
+		packs[root.pack] = PackageConfigs(null, [root.config], BitArray([true]));
+		long loop_counter = no_loop_limit ? long.max : 1_000_000;
+		constrainRec(root, packs, loop_counter);
+
+		// then use the latest configuration for each package as the final
+		// result
+		CONFIG[string] ret;
+		foreach (pack, pc; packs) {
+			if (pack == root.pack) continue;
+			foreach (i, c; pc.configs)
+				if (pc.included[i]) {
+					ret[pack] = c;
+					break;
 				}
+		}
 
-				foreach (c; getSpecificConfigs(basepack, ch))
-					if (!config.allConfigs.canFind(c))
-						config.allConfigs = c ~ config.allConfigs;
+		// remove any non-default optional dependencies
+		purgeOptionalDependencies(root, ret);
 
-				if (config.allConfigs.length > 0)
-					config.anyConfig = true;
+		return ret;
+	}
 
-				// store package depending on this for better error messages
-				config.origins ~= PackageConfigs.Depender(parent, ch);
+	protected abstract CONFIG[] getAllConfigs(string pack);
+	protected abstract CONFIG[] getSpecificConfigs(string pack, TreeNodes nodes);
+	protected abstract TreeNodes[] getChildren(TreeNode node);
+	protected abstract bool matches(CONFIGS configs, CONFIG config);
 
-				// eliminate configurations from which we know that they can't satisfy
-				// the uniquely defined root dependencies (==version or ~branch style dependencies)
-				if (parent_unique) config.allConfigs = config.allConfigs.filter!(c => matches(ch.configs, c)).array;
+	private static struct PackageConfigs
+	{
+		import std.bitmanip : BitArray;
 
-				configs[pidx] = config;
+		bool[string] visited; // key = sub package or empty string
+		CONFIG[] configs;
+		BitArray included;
```

Comment:
also the correlation between this length and configs.length could maybe be better represented by a struct containing both CONFIG and bool included, otherwise in the future someone might manipulate one array without knowing to change the other and somewhere else in the code a range error will occur.